clients/src/main/java/org/apache/kafka/common/record/FileRecords.java

```diff
@@ -135,17 +135,20 @@ public void readInto(ByteBuffer buffer, int position) throws IOException {
      * @return A sliced wrapper on this message set limited based on the given position and size
      */
     public FileRecords slice(int position, int size) throws IOException {
+        // Cache current size in case concurrent write changes it
+        int currentSizeInBytes = sizeInBytes();
+
         if (position < 0)
             throw new IllegalArgumentException("Invalid position: " + position + " in read from " + this);
-        if (position > sizeInBytes() - start)
+        if (position > currentSizeInBytes - start)
             throw new IllegalArgumentException("Slice from position " + position + " exceeds end position of " + this);
         if (size < 0)
             throw new IllegalArgumentException("Invalid size: " + size + " in read from " + this);
 
         int end = this.start + position + size;
         // handle integer overflow or if end is beyond the end of the file
-        if (end < 0 || end >= start + sizeInBytes())
-            end = start + sizeInBytes();
+        if (end < 0 || end > start + currentSizeInBytes)
```

Comment:
Was the boundary check wrong? You changed `>=` to `>`.

Reply:
The common case that we would see this is when the limit matches the file end exactly. So changing this to `>` would also have fixed this problem in the common case. I decided to remove it here though just because the equality check is redundant (if `end == start + currentSizeInBytes`, then there's no need to update it).

Reply:
Makes sense.